### PR TITLE
CAMEL-24634: camel-opa - new component to evaluate Open Policy Agent (Rego) policies against an Exchange

### PR DESCRIPTION
**Pull request by oscerd**
## What

Adds `camel-opa`, a new component that evaluates [Open Policy Agent](https://www.openpolicyagent.org/) (Rego) policies against an Exchange and records the allow/deny decision on it.

`camel-spiffe` (CAMEL-23305 / CAMEL-24571) gave Camel a way to establish *who* a caller is. Nothing answered *whether that caller is allowed*, so authorization logic ends up hand-coded in `.process()` / `.bean()` blocks — scattered across routes and impossible to change without redeploying the integration. `camel-shiro` and `camel-keycloak` each ship an `AuthorizationPolicy`, but both are bound to one identity provider and to coarse role/permission checks; neither supports externalised, attribute-based policy.

## Two usage modes

**Producer** — `opa:<policyPath>` evaluates the policy and sets `CamelOpaDecisionAllow` (Boolean) plus `CamelOpaDecision` (the raw decision document), leaving the body untouched, so the route decides with a filter:

```java
from("platform-http:/orders")
    .to("opa:authz/orders/allow")
    .filter(header(OpaConstants.DECISION_ALLOW).isEqualTo(true))
        .to("direct:handleOrder");
```

**Security policy** — `OpaSecurityPolicy implements AuthorizationPolicy`, so it wraps a route segment and a deny throws `CamelAuthorizationException`:

```java
from("platform-http:/orders")
    .policy(opaPolicy)
        .to("direct:handleOrder");
```

## Design decisions worth a look

- **Fails closed.** Any failure to reach a verdict — OPA unreachable, timeout, error, or a failure building/serializing the input document — throws: `OpaPolicyEvaluationException` from the producer, `CamelAuthorizationException` from the security policy. It never proceeds as allowed. `failOpen=true` reverses this and is marked `security = "insecure:dev"`.
- **A deny is distinct from a failure.** A policy that ran and said no is a decision; an unreachable PDP is not. The exception cause lets a route tell them apart.
- **The policy path is not header-overridable.** Taken from the endpoint only, so an inbound message cannot select which policy judges it.
- **Decision headers are written on every evaluation**, and are never sent back to OPA in the input document — so a verdict claimed by an inbound message neither survives into the route nor is shown to the policy.
- **The body is not sent by default** (`includeBody=false`); `includeHeaders` narrows the header set. Bodies can be large or streaming and most decisions need only headers. Values that are not JSON-native are stringified, and that conversion is deliberately **shallow** — a `Map` or `List` passes through as-is so a policy can read a JSON payload as a structure, with anything nested inside left to the SDK serializer.
- **Only a JSON boolean counts as an allow** — as the whole document, or as the `allowKey` entry of an object. Anything else is a deny, with the raw document kept on the exchange so the route can inspect deny reasons or obligations.
- **`OpaSecurityProcessor` extends `DelegateAsyncProcessor`**, following the `Policy` SPI's advice, so the wrapped route segment keeps the asynchronous routing engine. (`camel-shiro` and `camel-keycloak` both use the synchronous `DelegateProcessor`, which its own javadoc warns against; new code shouldn't repeat that.)

## Dependency

`com.styra:opa:2.1.1` — the official OPA Java SDK ([styrainc/opa-java](https://github.com/styrainc/opa-java)), Apache-2.0 and on Maven Central. It wraps OPA's REST Data API, which is the standard OPA sidecar deployment model. Its stray runtime-scoped `junit-jupiter-engine` is excluded.

In-process Rego evaluation via WASM (`com.styra.opa:opa-java-wasm`) is deliberately out of scope for this first increment — that artifact is still at 0.0.6.

## Testing

25 unit tests across producer behaviour, input-document shaping and the security policy, covering the fail-closed paths, the non-boolean decision shapes and the header-overwrite guarantees.

7 integration tests (`OpaIT`) against a real OPA server, via a new `camel-test-infra-opa` service: producer allow/deny, a decision object read through `allowKey`, the deny reasons alongside it, fail-closed on an unreachable server, and the security policy in both directions. The policy they run, `authz.rego`, is written as a worked example of the input document and both decision shapes, and the component docs link to it.

`openpolicyagent/opa` is published only for amd64 and arm64, so the ITs carry `skipITs.ppc64le` / `skipITs.s390x` (as `camel-cyberark-vault` does for conjur) and pull through `mirror.gcr.io`. The license plugin gained a `.rego` comment-style mapping.

Full reactor build green; `camel-catalog` tests green.

## Review

`gnodet`'s round (non-blocking) is addressed in 35f7c3e: the shallow-conversion contract is now spelled out in `toJsonSafe`'s javadoc and in `opa-component.adoc`. The `doStop()` null-assignment observation and the two `ast-grep` broad-catch hits are answered in the threads — kept as-is, with the reasoning recorded there.

`Croway`'s request for e2e tests that double as documentation is addressed in 7175de5 (see the Testing section).

`main` only — new component, nothing to backport. No upgrade-guide entry: the guide is for migration, and this adds nothing existing users must migrate.

_Claude Code on behalf of @oscerd_

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by gnodet — COMMENTED** [review]
Good work. This is a well-designed component with thorough security defaults (fail-closed, policy path not header-overridable, decision headers always overwritten, `DelegateAsyncProcessor` over `DelegateProcessor`). The test coverage is strong — 25 tests across producer, input shaping, and the security policy, hitting all the edge cases that matter: fail-closed, fail-open, non-boolean verdicts, forged headers, custom allow keys.

A few observations below. Nothing blocking.

Static analysis (`ast-grep`) flagged the `catch (Exception e)` blocks in `OpaPolicyEvaluator.evaluate()` and `OpaSecurityProcessor.process()` as overly broad. Both are intentional — the evaluator's catch is the fail-closed implementation (deliberately catching both `OPAException` and serialization failures), and the processor's catch is the standard Camel async processor pattern. No action needed.

_This review was generated by an AI agent, Hermès on behalf of @gnodet._

File: components/camel-opa/src/main/java/org/apache/camel/component/opa/OpaEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                           + " The path is taken from the endpoint only: it is deliberately not overridable by a"
+                           + " message header, so that an inbound message cannot select which policy judges it.")
+    @Metadata(required = true)
+    private String policyPath;
+
+    @UriParam
+    private OpaConfiguration configuration;
+
+    private OPAClient opaClient;
+    private volatile OpaPolicyEvaluator evaluator;
```
Comment (gnodet) [inline components/camel-opa/src/main/java/org/apache/camel/component/opa/OpaEndpoint.java:50]: 💡 The `evaluator` field is `volatile` here and in `OpaSecurityPolicy`, which ensures visibility. Note that `doStop()` sets this to `null` before calling `super.doStop()` — in theory a racing `process()` call could see `null` and NPE. In practice this is safe because Camel's lifecycle stops the producer (draining in-flight exchanges) before stopping the endpoint. Just worth being aware of if the ordering ever changes.

File: components/camel-opa/src/main/java/org/apache/camel/component/opa/OpaPolicyEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+    /**
+     * Builds the {@code input} document handed to OPA.
+     */
+    protected Map<String, Object> buildInput(Exchange exchange) {
+        Map<String, Object> input = new LinkedHashMap<>();
+        Map<String, Object> headers = new LinkedHashMap<>();
+        for (Map.Entry<String, Object> entry : exchange.getMessage().getHeaders().entrySet()) {
+            String name = entry.getKey();
+            // never feed our own decision headers back in: a policy must not be able to read a verdict
```
Comment (gnodet) [inline components/camel-opa/src/main/java/org/apache/camel/component/opa/OpaPolicyEvaluator.java:112]: 💡 `toJsonSafe` passes `Map` and `List` through as-is. If a header value is a `Map` containing non-JSON-native nested values (e.g., a `Map<String, Date>`), those nested values will be serialized by the OPA SDK and may produce unexpected JSON. This is fine for the common case (headers are mostly strings/numbers), but worth a note in the Javadoc if you want to be precise about the contract.

File: components/camel-opa/src/main/java/org/apache/camel/component/opa/OpaEndpoint.java (hunk 1 above)
Comment (oscerd) [reply components/camel-opa/src/main/java/org/apache/camel/component/opa/OpaEndpoint.java:50]: Good catch on the ordering, and thanks for checking the lifecycle rather than just flagging the shape.

Leaving it as-is, for two reasons. Camel's shutdown stops the routes and drains their producers before it stops the endpoints, so `process()` and `doStop()` do not overlap — and `camel-spiffe` nulls its `WorkloadApiClient` in exactly the same place, so this at least stays consistent with the neighbouring component.

The nulling is also load-bearing rather than tidiness: `OPAClient` holds an `OpaApiClient` -> `OPAHTTPClient` -> `java.net.http.HttpClient`, which is not closeable on 17 and whose selector thread only goes away once the client is unreachable. Dropping the reference is what lets that happen when a stopped endpoint is retained in the registry. If I removed the assignments to close the theoretical window, a racing exchange would get a stale-but-working evaluator (fine in itself) at the cost of pinning that thread.

Worth revisiting if the shutdown ordering ever changes, as you say — happy to add a null-guard in the producer if you would rather have belt and braces.

_Claude Code on behalf of @oscerd_

File: components/camel-opa/src/main/java/org/apache/camel/component/opa/OpaPolicyEvaluator.java (hunk 2 above)
Comment (oscerd) [reply components/camel-opa/src/main/java/org/apache/camel/component/opa/OpaPolicyEvaluator.java:112]: Agreed, the javadoc was overstating the contract. Fixed in 35f7c3e.

`toJsonSafe` now says explicitly that the conversion is shallow — a `Map` or `List` is passed through as-is, so anything non-JSON-native nested inside it is left to the SDK's serializer, and a policy reading nested structures should not assume the same string conversion applies at depth. Added the same note to `opa-component.adoc`, since that is where a policy author is more likely to look.

Left the behaviour alone: deep-converting would mean walking arbitrary structures on every exchange, and for the common case you describe (a `Map` body, which is exactly what a JSON payload converts to) passing it through is what makes the policy able to read it as a structure at all.

_Claude Code on behalf of @oscerd_

**Comment by oscerd** [comment]
Thanks for the review.

On the `ast-grep` hit for the two broad catches — your reading is right, and both are deliberate:

- `OpaPolicyEvaluator.evaluate()` catches `Exception` rather than `OPAException` on purpose. Fail-closed has to hold regardless of which layer gave up, so a Jackson failure serializing the input document, or a type-converter failure building it, has to land in the same place as an unreachable OPA server. Narrowing it to `OPAException` would make an input-serialization bug bypass `failOpen` handling and surface as an unhandled runtime exception instead. There is a test for exactly that (`failsClosedWhenTheSdkFailsWithSomethingOtherThanAnOpaException`).
- `OpaSecurityProcessor.process()` is the standard `DelegateAsyncProcessor` contract — set the exception on the exchange, call the callback, return true — so nothing to narrow there either.

Addressed your javadoc point in 35f7c3e and replied on the `doStop()` thread with the reasoning for keeping the assignments. Both threads left open for you to resolve.

_Claude Code on behalf of @oscerd_

**Comment by Croway** [comment]
@oscerd would it be possible to have e2e integration tests? that might be used as documentation too

**Review by gnodet — APPROVED** [review]
Re-review after commits 35f7c3e and 47a405c.

**Previous findings status:**
- `toJsonSafe` shallow-conversion documentation — **Addressed.** Javadoc now clearly spells out the shallow contract (Map/List pass-through, nested non-JSON-native values left to SDK serializer). Same note added to `opa-component.adoc`. Both accurate and consistent.
- `doStop()` null-assignment observation — **Acknowledged.** Author's reasoning is sound: Camel lifecycle drains producers before stopping endpoints, and the null-assignment is load-bearing (drops the reference to the `HttpClient` selector thread). Consistent with `camel-spiffe`.
- `ast-grep` broad-catch hits — **Acknowledged.** Both `catch (Exception e)` blocks are intentional (fail-closed semantics + standard `DelegateAsyncProcessor` contract). Tests confirm the behavior.

No new issues in the follow-up commits. The generated catalog files in 47a405c are consistent. Clean to merge.

_This review was generated by an AI agent, Hermès on behalf of @gnodet._

**Comment by oscerd** [comment]
@Croway good call — the unit tests all mock the OPA client, so nothing was exercising real Rego evaluation over the wire. Added in 7175de5.

`camel-test-infra-opa` starts an OPA server with an empty policy set, so the service stays independent of any particular policy and tests load their own through the OPA REST API. `OpaIT` then drives the component end to end against it: producer allow and deny, a decision object read through `allowKey`, the deny reasons that come back with it, fail-closed against an unreachable server, and the security policy in both directions. Seven tests, green locally.

On your second point — `authz.rego` is deliberately written as documentation rather than as bare test fixture. It spells out the shape of the input document Camel sends, shows why every decision rule needs a `default` (an undefined decision comes back from OPA as an evaluation error, not a deny, so the component fails closed on it), and gives both decision shapes: a plain boolean and an object carrying reasons. The component docs now link to it, with `OpaIT` next to it as the matching routes.

One caveat worth flagging: `openpolicyagent/opa` only publishes amd64 and arm64, so the ITs carry `skipITs.ppc64le` / `skipITs.s390x`, same as `camel-cyberark-vault` does for the conjur image. Pulled via `mirror.gcr.io` rather than Docker Hub.

_Claude Code on behalf of @oscerd_

**Review by gnodet — APPROVED** [review]
Re-review after commit 7175de5 (integration tests against a real OPA server).

**Previous findings status:**
All findings from the initial review were acknowledged/addressed in 35f7c3e (javadoc for shallow conversion contract) — confirmed in the prior re-review. No outstanding items.

**Review of the new commit (7175de5):**

This addresses @Croway's request for e2e integration tests. The implementation is solid:

- `camel-test-infra-opa` follows the established test-infra pattern exactly (infra service interface, local container, remote, singleton factory) — consistent with `camel-test-infra-neo4j`, `camel-test-infra-pinecone`, etc.
- `OpaIT` covers the right surfaces: producer allow/deny, decision object with `allowKey`, deny reasons, fail-closed on unreachable server, security policy allow/deny. Seven tests, all exercising real Rego evaluation over the wire.
- `authz.rego` is well-documented as a worked example — doubles as documentation as requested.
- `skipITs.ppc64le`/`skipITs.s390x` for the OPA image (amd64/arm64 only) follows the `camel-cyberark-vault` precedent.
- `mirror.gcr.io` pull avoids Docker Hub rate limits — matches project convention.
- License plugin `.rego` mapping added correctly.
- The `@BeforeAll uploadPolicy()` approach (loading policy via REST API into a clean OPA server) keeps the test-infra service independent of any particular policy — clean design.

No issues found. Clean to merge.

_This review was generated by an AI agent, Hermès on behalf of @gnodet._

**Review by davsclaus — APPROVED** [review]

**Review by Croway — APPROVED** [review]

**Comment by andreaTP** [comment]
Have you considered evaluating OPA rules in-process?
I'm maintaining this repo since some time: https://github.com/StyraOSS/opa-java-wasm
and seems to be used enough despite the low number of stars.

**Comment by oscerd** [comment]
@andreaTP can you elaborate a bit? I didn't know about that project.

**Comment by oscerd** [comment]
This is the first version of the component and it's not released. So we are not tied to any library

**Comment by andreaTP** [comment]
@oscerd sure!
The `opa` cli can directly compile `.rego` rules to `wasm` and you can evaluate them inside any Java application(in this case Camel). ref: https://www.openpolicyagent.org/docs/wasm

This removes the requirement of having an OPA server available(less infrastructure, no need for an external container etc.) while preserving all the original semantics of the Rego rule.

Additionally, those rules run in a safe Wasm sandbox where you can easily control the used memory and/or timeout execution when necessary.
